{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np # linear algebra\n",
    "import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n",
    "import glob\n",
    "import json\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('ggplot')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>Microsoft Academic Paper ID</th>\n",
       "      <th>WHO #Covidence</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>xqhn0vbp</td>\n",
       "      <td>1e1286db212100993d03cc22374b624f7caee956</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Airborne rhinovirus detection and effect of ul...</td>\n",
       "      <td>10.1186/1471-2458-3-5</td>\n",
       "      <td>PMC140314</td>\n",
       "      <td>12525263.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: Rhinovirus, the most common cause ...</td>\n",
       "      <td>2003-01-13</td>\n",
       "      <td>Myatt, Theodore A; Johnston, Sebastian L; Rudn...</td>\n",
       "      <td>BMC Public Health</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>gi6uaa83</td>\n",
       "      <td>8ae137c8da1607b3a8e4c946c07ca8bda67f88ac</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Discovering human history from stomach bacteria</td>\n",
       "      <td>10.1186/gb-2003-4-5-213</td>\n",
       "      <td>PMC156578</td>\n",
       "      <td>12734001.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Recent analyses of human pathogens have reveal...</td>\n",
       "      <td>2003-04-28</td>\n",
       "      <td>Disotell, Todd R</td>\n",
       "      <td>Genome Biol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>le0ogx1s</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMC</td>\n",
       "      <td>A new recruit for the army of the men of death</td>\n",
       "      <td>10.1186/gb-2003-4-7-113</td>\n",
       "      <td>PMC193621</td>\n",
       "      <td>12844350.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>The army of the men of death, in John Bunyan's...</td>\n",
       "      <td>2003-06-27</td>\n",
       "      <td>Petsko, Gregory A</td>\n",
       "      <td>Genome Biol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>fy4w7xz8</td>\n",
       "      <td>0104f6ceccf92ae8567a0102f89cbb976969a774</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Association of HLA class I with severe acute r...</td>\n",
       "      <td>10.1186/1471-2350-4-9</td>\n",
       "      <td>PMC212558</td>\n",
       "      <td>12969506.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: The human leukocyte antigen (HLA) ...</td>\n",
       "      <td>2003-09-12</td>\n",
       "      <td>Lin, Marie; Tseng, Hsiang-Kuang; Trejaut, Jean...</td>\n",
       "      <td>BMC Med Genet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0qaoam29</td>\n",
       "      <td>5b68a553a7cbbea13472721cd1ad617d42b40c26</td>\n",
       "      <td>PMC</td>\n",
       "      <td>A double epidemic model for the SARS propagation</td>\n",
       "      <td>10.1186/1471-2334-3-19</td>\n",
       "      <td>PMC222908</td>\n",
       "      <td>12964944.0</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: An epidemic of a Severe Acute Resp...</td>\n",
       "      <td>2003-09-10</td>\n",
       "      <td>Ng, Tuen Wai; Turinici, Gabriel; Danchin, Antoine</td>\n",
       "      <td>BMC Infect Dis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51073</td>\n",
       "      <td>4xhcamks</td>\n",
       "      <td>e14e7f6ffed1a74954167e4fb52d1b2f156171cd</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Prevalence of antibodies against transmissible...</td>\n",
       "      <td>10.1016/j.vas.2018.11.003</td>\n",
       "      <td>PMC7104163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract Transmissible gastroenteritis (TGE) i...</td>\n",
       "      <td>2019-06-30</td>\n",
       "      <td>Valkó, Anna; Bálint, Ádám; Bozsa, Ágnes; Cságo...</td>\n",
       "      <td>Veterinary and Animal Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.vas.2018.11.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51074</td>\n",
       "      <td>z2rqz3jx</td>\n",
       "      <td>9550137f5e105d2dbc14d91659a8c3cdd815c9b1</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Can we increase public awareness without creat...</td>\n",
       "      <td>10.1016/j.pec.2013.10.023</td>\n",
       "      <td>PMC7119035</td>\n",
       "      <td>24284163.0</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-02-28</td>\n",
       "      <td>Al Turki, Yousef Abdullah</td>\n",
       "      <td>Patient Education and Counseling</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.pec.2013.10.023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51075</td>\n",
       "      <td>wncuc903</td>\n",
       "      <td>2b3a3e3956afd65a8fa11d3867da3e58f8915514</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Fast degrading polyesters as siRNA nano-carrie...</td>\n",
       "      <td>10.1016/j.jconrel.2008.06.010</td>\n",
       "      <td>PMC7125568</td>\n",
       "      <td>18619502.0</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract A potential siRNA carrier for pulmona...</td>\n",
       "      <td>2008-12-18</td>\n",
       "      <td>Nguyen, Juliane; Steele, Terry W.J.; Merkel, O...</td>\n",
       "      <td>Journal of Controlled Release</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.jconrel.2008.06.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51076</td>\n",
       "      <td>4x1ckr2i</td>\n",
       "      <td>a67dae6e5f2b58a6f16aaed4011c694f9e64dd88</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>eIF3f: A central regulator of the antagonism a...</td>\n",
       "      <td>10.1016/j.biocel.2013.06.001</td>\n",
       "      <td>PMC7108353</td>\n",
       "      <td>23769948.0</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract The eukaryotic initiation factor 3 su...</td>\n",
       "      <td>2013-10-31</td>\n",
       "      <td>Sanchez, Anthony M.J.; Csibi, Alfredo; Raibon,...</td>\n",
       "      <td>The International Journal of Biochemistry &amp; Ce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.biocel.2013.06.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51077</td>\n",
       "      <td>1peoakxt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>The pathogenic role of virus-specific antibody...</td>\n",
       "      <td>10.1016/0165-5728(91)91228-5</td>\n",
       "      <td>PMC7119703</td>\n",
       "      <td>NaN</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1991-12-31</td>\n",
       "      <td>Schwender, S.; Imrich, H.; Dörries, R.</td>\n",
       "      <td>Journal of Neuroimmunology</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0165-5728(91)91228-5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51078 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cord_uid                                       sha  source_x  \\\n",
       "0      xqhn0vbp  1e1286db212100993d03cc22374b624f7caee956       PMC   \n",
       "1      gi6uaa83  8ae137c8da1607b3a8e4c946c07ca8bda67f88ac       PMC   \n",
       "2      le0ogx1s                                       NaN       PMC   \n",
       "3      fy4w7xz8  0104f6ceccf92ae8567a0102f89cbb976969a774       PMC   \n",
       "4      0qaoam29  5b68a553a7cbbea13472721cd1ad617d42b40c26       PMC   \n",
       "...         ...                                       ...       ...   \n",
       "51073  4xhcamks  e14e7f6ffed1a74954167e4fb52d1b2f156171cd  Elsevier   \n",
       "51074  z2rqz3jx  9550137f5e105d2dbc14d91659a8c3cdd815c9b1  Elsevier   \n",
       "51075  wncuc903  2b3a3e3956afd65a8fa11d3867da3e58f8915514  Elsevier   \n",
       "51076  4x1ckr2i  a67dae6e5f2b58a6f16aaed4011c694f9e64dd88  Elsevier   \n",
       "51077  1peoakxt                                       NaN  Elsevier   \n",
       "\n",
       "                                                   title  \\\n",
       "0      Airborne rhinovirus detection and effect of ul...   \n",
       "1        Discovering human history from stomach bacteria   \n",
       "2         A new recruit for the army of the men of death   \n",
       "3      Association of HLA class I with severe acute r...   \n",
       "4       A double epidemic model for the SARS propagation   \n",
       "...                                                  ...   \n",
       "51073  Prevalence of antibodies against transmissible...   \n",
       "51074  Can we increase public awareness without creat...   \n",
       "51075  Fast degrading polyesters as siRNA nano-carrie...   \n",
       "51076  eIF3f: A central regulator of the antagonism a...   \n",
       "51077  The pathogenic role of virus-specific antibody...   \n",
       "\n",
       "                                 doi       pmcid   pubmed_id    license  \\\n",
       "0              10.1186/1471-2458-3-5   PMC140314  12525263.0      no-cc   \n",
       "1            10.1186/gb-2003-4-5-213   PMC156578  12734001.0      no-cc   \n",
       "2            10.1186/gb-2003-4-7-113   PMC193621  12844350.0      no-cc   \n",
       "3              10.1186/1471-2350-4-9   PMC212558  12969506.0      no-cc   \n",
       "4             10.1186/1471-2334-3-19   PMC222908  12964944.0      no-cc   \n",
       "...                              ...         ...         ...        ...   \n",
       "51073      10.1016/j.vas.2018.11.003  PMC7104163         NaN  els-covid   \n",
       "51074      10.1016/j.pec.2013.10.023  PMC7119035  24284163.0  els-covid   \n",
       "51075  10.1016/j.jconrel.2008.06.010  PMC7125568  18619502.0  els-covid   \n",
       "51076   10.1016/j.biocel.2013.06.001  PMC7108353  23769948.0  els-covid   \n",
       "51077   10.1016/0165-5728(91)91228-5  PMC7119703         NaN  els-covid   \n",
       "\n",
       "                                                abstract publish_time  \\\n",
       "0      BACKGROUND: Rhinovirus, the most common cause ...   2003-01-13   \n",
       "1      Recent analyses of human pathogens have reveal...   2003-04-28   \n",
       "2      The army of the men of death, in John Bunyan's...   2003-06-27   \n",
       "3      BACKGROUND: The human leukocyte antigen (HLA) ...   2003-09-12   \n",
       "4      BACKGROUND: An epidemic of a Severe Acute Resp...   2003-09-10   \n",
       "...                                                  ...          ...   \n",
       "51073  Abstract Transmissible gastroenteritis (TGE) i...   2019-06-30   \n",
       "51074                                                NaN   2014-02-28   \n",
       "51075  Abstract A potential siRNA carrier for pulmona...   2008-12-18   \n",
       "51076  Abstract The eukaryotic initiation factor 3 su...   2013-10-31   \n",
       "51077                                                NaN   1991-12-31   \n",
       "\n",
       "                                                 authors  \\\n",
       "0      Myatt, Theodore A; Johnston, Sebastian L; Rudn...   \n",
       "1                                       Disotell, Todd R   \n",
       "2                                      Petsko, Gregory A   \n",
       "3      Lin, Marie; Tseng, Hsiang-Kuang; Trejaut, Jean...   \n",
       "4      Ng, Tuen Wai; Turinici, Gabriel; Danchin, Antoine   \n",
       "...                                                  ...   \n",
       "51073  Valkó, Anna; Bálint, Ádám; Bozsa, Ágnes; Cságo...   \n",
       "51074                          Al Turki, Yousef Abdullah   \n",
       "51075  Nguyen, Juliane; Steele, Terry W.J.; Merkel, O...   \n",
       "51076  Sanchez, Anthony M.J.; Csibi, Alfredo; Raibon,...   \n",
       "51077             Schwender, S.; Imrich, H.; Dörries, R.   \n",
       "\n",
       "                                                 journal  \\\n",
       "0                                      BMC Public Health   \n",
       "1                                            Genome Biol   \n",
       "2                                            Genome Biol   \n",
       "3                                          BMC Med Genet   \n",
       "4                                         BMC Infect Dis   \n",
       "...                                                  ...   \n",
       "51073                      Veterinary and Animal Science   \n",
       "51074                   Patient Education and Counseling   \n",
       "51075                      Journal of Controlled Release   \n",
       "51076  The International Journal of Biochemistry & Ce...   \n",
       "51077                         Journal of Neuroimmunology   \n",
       "\n",
       "       Microsoft Academic Paper ID WHO #Covidence  has_pdf_parse  \\\n",
       "0                              NaN            NaN           True   \n",
       "1                              NaN            NaN           True   \n",
       "2                              NaN            NaN          False   \n",
       "3                              NaN            NaN           True   \n",
       "4                              NaN            NaN           True   \n",
       "...                            ...            ...            ...   \n",
       "51073                          NaN            NaN           True   \n",
       "51074                          NaN            NaN           True   \n",
       "51075                          NaN            NaN           True   \n",
       "51076                          NaN            NaN           True   \n",
       "51077                          NaN            NaN          False   \n",
       "\n",
       "       has_pmc_xml_parse  full_text_file  \\\n",
       "0                   True  custom_license   \n",
       "1                   True  custom_license   \n",
       "2                   True  custom_license   \n",
       "3                   True  custom_license   \n",
       "4                   True  custom_license   \n",
       "...                  ...             ...   \n",
       "51073              False  custom_license   \n",
       "51074              False  custom_license   \n",
       "51075              False  custom_license   \n",
       "51076              False  custom_license   \n",
       "51077              False  custom_license   \n",
       "\n",
       "                                                     url  \n",
       "0      https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "1      https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "2      https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "3      https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  \n",
       "4      https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  \n",
       "...                                                  ...  \n",
       "51073          https://doi.org/10.1016/j.vas.2018.11.003  \n",
       "51074          https://doi.org/10.1016/j.pec.2013.10.023  \n",
       "51075      https://doi.org/10.1016/j.jconrel.2008.06.010  \n",
       "51076       https://doi.org/10.1016/j.biocel.2013.06.001  \n",
       "51077       https://doi.org/10.1016/0165-5728(91)91228-5  \n",
       "\n",
       "[51078 rows x 18 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_path = f'CORD-19-research-challenge/metadata.csv'\n",
    "meta_df = pd.read_csv(metadata_path, sep=',')\n",
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 51078 entries, 0 to 51077\n",
      "Data columns (total 18 columns):\n",
      "cord_uid                       51078 non-null object\n",
      "sha                            38022 non-null object\n",
      "source_x                       51078 non-null object\n",
      "title                          50920 non-null object\n",
      "doi                            47741 non-null object\n",
      "pmcid                          41082 non-null object\n",
      "pubmed_id                      37861 non-null float64\n",
      "license                        51078 non-null object\n",
      "abstract                       42352 non-null object\n",
      "publish_time                   51070 non-null object\n",
      "authors                        48891 non-null object\n",
      "journal                        46368 non-null object\n",
      "Microsoft Academic Paper ID    964 non-null float64\n",
      "WHO #Covidence                 1768 non-null object\n",
      "has_pdf_parse                  51078 non-null bool\n",
      "has_pmc_xml_parse              51078 non-null bool\n",
      "full_text_file                 42511 non-null object\n",
      "url                            50776 non-null object\n",
      "dtypes: bool(2), float64(2), object(14)\n",
      "memory usage: 6.3+ MB\n"
     ]
    }
   ],
   "source": [
    "meta_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_name(author):\n",
    "    middle_name = \" \".join(author['middle'])\n",
    "    \n",
    "    if author['middle']:\n",
    "        return \" \".join([author['first'], middle_name, author['last']])\n",
    "    else:\n",
    "        return \" \".join([author['first'], author['last']])\n",
    "\n",
    "def format_affiliation(affiliation):\n",
    "    text = []\n",
    "    location = affiliation.get('location')\n",
    "    if location:\n",
    "        text.extend(list(affiliation['location'].values()))\n",
    "    \n",
    "    institution = affiliation.get('institution')\n",
    "    if institution:\n",
    "        text = [institution] + text\n",
    "    return \", \".join(text)\n",
    "\n",
    "def format_authors(authors, with_affiliation=False):\n",
    "    name_ls = []\n",
    "    \n",
    "    for author in authors:\n",
    "        name = format_name(author)\n",
    "        if with_affiliation:\n",
    "            affiliation = format_affiliation(author['affiliation'])\n",
    "            if affiliation:\n",
    "                name_ls.append(f\"{name} ({affiliation})\")\n",
    "            else:\n",
    "                name_ls.append(name)\n",
    "        else:\n",
    "            name_ls.append(name)\n",
    "    \n",
    "    return \", \".join(name_ls)\n",
    "\n",
    "def format_body(body_text):\n",
    "    texts = [(di['section'], di['text']) for di in body_text]\n",
    "    texts_di = {di['section']: \"\" for di in body_text}\n",
    "    \n",
    "    for section, text in texts:\n",
    "        texts_di[section] += text\n",
    "\n",
    "    body = \"\"\n",
    "\n",
    "    for section, text in texts_di.items():\n",
    "        body += section\n",
    "        body += \"\\n\\n\"\n",
    "        body += text\n",
    "        body += \"\\n\\n\"\n",
    "    \n",
    "    return body\n",
    "\n",
    "def format_bib(bibs):\n",
    "    if type(bibs) == dict:\n",
    "        bibs = list(bibs.values())\n",
    "    bibs = deepcopy(bibs)\n",
    "    formatted = []\n",
    "    \n",
    "    for bib in bibs:\n",
    "        bib['authors'] = format_authors(\n",
    "            bib['authors'], \n",
    "            with_affiliation=False\n",
    "        )\n",
    "        formatted_ls = [str(bib[k]) for k in ['title', 'authors', 'venue', 'year']]\n",
    "        formatted.append(\", \".join(formatted_ls))\n",
    "\n",
    "    return \"; \".join(formatted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_files(dirname):\n",
    "    filenames = os.listdir(dirname)\n",
    "    raw_files = []\n",
    "\n",
    "    for filename in tqdm(filenames):\n",
    "        filename = dirname + filename\n",
    "        file = json.load(open(filename, 'rb'))\n",
    "        raw_files.append(file)\n",
    "    \n",
    "    return raw_files\n",
    "\n",
    "def generate_clean_df(all_files):\n",
    "    cleaned_files = []\n",
    "    \n",
    "    for file in tqdm(all_files):\n",
    "        features = [\n",
    "            file['paper_id'],\n",
    "            file['metadata']['title'],\n",
    "            format_authors(file['metadata']['authors']),\n",
    "            format_authors(file['metadata']['authors'], \n",
    "                           with_affiliation=True),\n",
    "            format_body(file['abstract']),\n",
    "            format_body(file['body_text']),\n",
    "            format_bib(file['bib_entries']),\n",
    "            file['metadata']['authors'],\n",
    "            file['bib_entries']\n",
    "        ]\n",
    "\n",
    "        cleaned_files.append(features)\n",
    "\n",
    "    col_names = ['paper_id', 'title', 'authors',\n",
    "                 'affiliations', 'abstract', 'text', \n",
    "                 'bibliography','raw_authors','raw_bibliography']\n",
    "\n",
    "    clean_df = pd.DataFrame(cleaned_files, columns=col_names)\n",
    "    clean_df.head()\n",
    "    \n",
    "    return clean_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biorxiv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pprint import pprint\n",
    "from copy import deepcopy\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of articles retrieved from biorxiv: 1625\n"
     ]
    }
   ],
   "source": [
    "biorxiv_dir = 'CORD-19-research-challenge/biorxiv_medrxiv/biorxiv_medrxiv/pdf_json/'\n",
    "filenames = os.listdir(biorxiv_dir)\n",
    "print(\"Number of articles retrieved from biorxiv:\", len(filenames))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_files = []\n",
    "\n",
    "for filename in filenames:\n",
    "    filename = biorxiv_dir + filename\n",
    "    file = json.load(open(filename, 'rb'))\n",
    "    all_files.append(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dictionary keys: dict_keys(['paper_id', 'metadata', 'abstract', 'body_text', 'bib_entries', 'ref_entries', 'back_matter'])\n"
     ]
    }
   ],
   "source": [
    "file = all_files[0]\n",
    "print(\"Dictionary keys:\", file.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'word count: 194 22 Text word count: 5168 23 24 25 author/funder. '\n",
      "          'All rights reserved. No reuse allowed without permission. Abstract '\n",
      "          '27 The positive stranded RNA genomes of picornaviruses comprise a '\n",
      "          'single large open reading 28 frame flanked by 5′ and 3′ '\n",
      "          'untranslated regions (UTRs). Foot-and-mouth disease virus (FMDV) 29 '\n",
      "          'has an unusually large 5′ UTR (1.3 kb) containing five structural '\n",
      "          'domains. These include the 30 internal ribosome entry site (IRES), '\n",
      "          'which facilitates initiation of translation, and the cis-acting 31 '\n",
      "          'replication element (cre). Less well characterised structures are a '\n",
      "          '5′ terminal 360 nucleotide 32 stem-loop, a variable length '\n",
      "          'poly-C-tract of approximately 100-200 nucleotides and a series of '\n",
      "          '33 two to four tandemly repeated pseudoknots (PKs). We investigated '\n",
      "          'the structures of the PKs 34 by selective 2′ hydroxyl acetylation '\n",
      "          'analysed by primer extension (SHAPE) analysis and 35 determined '\n",
      "          'their contribution to genome replication by mutation and deletion '\n",
      "          'experiments. 36 SHAPE and mutation experiments confirmed the '\n",
      "          'importance of the previously predicted PK 37 structures for their '\n",
      "          'function. Deletion experiments showed that although PKs are not '\n",
      "          'essential 38'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'for replication, they provide genomes with a competitive advantage. '\n",
      "          'However, although 39 replicons and full-length genomes lacking all '\n",
      "          'PKs were replication competent, no infectious 40 virus was rescued '\n",
      "          'from genomes containing less than one PK copy. This is consistent '\n",
      "          'with our 41 earlier report describing the presence of putative '\n",
      "          'packaging signals in the PK region. 42 43 author/funder. All rights '\n",
      "          'reserved. No reuse allowed without permission.'}]\n"
     ]
    }
   ],
   "source": [
    "# Biorxiv: Abstract\n",
    "pprint(file['abstract'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "body_text type: <class 'list'>\n",
      "body_text length: 20\n",
      "body_text keys: dict_keys(['text', 'cite_spans', 'ref_spans', 'section'])\n"
     ]
    }
   ],
   "source": [
    "# Biorxiv: body text\n",
    "print(\"body_text type:\", type(file['body_text']))\n",
    "print(\"body_text length:\", len(file['body_text']))\n",
    "print(\"body_text keys:\", file['body_text'][0].keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "body_text content:\n",
      "[{'cite_spans': [],\n",
      "  'ref_spans': [{...}],\n",
      "  'section': '',\n",
      "  'text': 'VP3, and VP0 (which is further processed to VP2 and VP4 during '\n",
      "          'virus assembly) (6). The P2 64 and P3 regions encode the '\n",
      "          'non-structural proteins 2B and 2C and 3A, 3B (1-3) (VPg), 3C pro '\n",
      "          'and 4 structural protein-coding region is replaced by reporter '\n",
      "          'genes, allow the study of genome 68 replication without the '\n",
      "          'requirement for high containment (9, 10) ( figure 1A ).'},\n",
      " {'cite_spans': [{...}, {...}, {...}, {...}, {...}, {...}],\n",
      "  'ref_spans': [],\n",
      "  'section': '70',\n",
      "  'text': 'The FMDV 5′ UTR is the largest known picornavirus UTR, comprising '\n",
      "          'approximately 1300 71 nucleotides and containing several highly '\n",
      "          'structured regions. The first 360 nucleotides at the 5′ 72 end are '\n",
      "          'predicted to fold into a single large stem loop termed the '\n",
      "          'S-fragment, followed by a The PKs were originally predicted in 1987 '\n",
      "          'and consist of two to four tandem repeats of a ~48 86 nucleotide '\n",
      "          'region containing a small stem loop and downstream interaction site '\n",
      "          '(figure 1B) 87 (12). Due to the sequence similarity between the PKs '\n",
      "          '(figure 1C), it is speculated that they 88 were formed by '\n",
      "          'duplication events during viral replication, probably involving '\n",
      "          'recombination. 89 Between two and four PKs are present in different '\n",
      "          'virus isolates but no strain has been 90 identified with less than '\n",
      "          'two PKs, emphasising their potential importance in the viral life '\n",
      "          'cycle 91 (19, 20) . The presence of PKs has been reported in the 5′ '\n",
      "          'UTR of other picornaviruses such as 92 author/funder. All rights '\n",
      "          'reserved. No reuse allowed without permission. can occur in the '\n",
      "          'absence of PKs at least one is required for wild-type (wt) '\n",
      "          'replication. 104 Furthermore, competition experiments showed that '\n",
      "          'extra copies of PKs conferred a replicative 105 advantage to '\n",
      "          'genomes. Although replicons and full-length genomes lacking PKs '\n",
      "          'were 106 replication-competent, no infectious virus was rescued '\n",
      "          'from genomes containing less than one 107 PK copy. This is '\n",
      "          'consistent with our earlier report describing the presence of '\n",
      "          'putative 108 packaging signals in the PK region (22). 109 110 '\n",
      "          'author/funder. All rights reserved. No reuse allowed without '\n",
      "          'permission. Plasmid construction. 117 The FMDV replicon plasmids, '\n",
      "          'pRep-ptGFP, and the replication-defective polymerase mutant 118 '\n",
      "          'control, 3D-GNN, have already been described (10).'}]\n"
     ]
    }
   ],
   "source": [
    "print(\"body_text content:\")\n",
    "pprint(file['body_text'][:2], depth=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['',\n",
      " '70',\n",
      " '120',\n",
      " '135',\n",
      " '136',\n",
      " '144',\n",
      " '301',\n",
      " 'Function of the PKs in replication is dependent on downstream interactions '\n",
      " 'and 350',\n",
      " '368',\n",
      " '468',\n",
      " '479']\n"
     ]
    }
   ],
   "source": [
    "texts = [(di['section'], di['text']) for di in file['body_text']]\n",
    "texts_di = {di['section']: \"\" for di in file['body_text']}\n",
    "for section, text in texts:\n",
    "    texts_di[section] += text\n",
    "\n",
    "pprint(list(texts_di.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "VP3, and VP0 (which is further processed to VP2 and VP4 during virus assembly) (6). The P2 64 and P3 regions encode the non-structural proteins 2B and 2C and 3A, 3B (1-3) (VPg), 3C pro and 4 structural protein-coding region is replaced by reporter genes, allow the study of genome 68 replication without the requirement for high containment (9, 10) ( figure 1A ).\n",
      "\n",
      "70\n",
      "\n",
      "The FMDV 5′ UTR is the largest known picornavirus UTR, comprising approximately 1300 71 nucleotides and containing several highly structured regions. The first 360 nucleotides at the 5′ 72 end are predicted to fold into a single large stem loop termed the S-fragment, followed by a The PKs were originally predicted in 1987 and consist of two to four tandem repeats of a ~48 86 nucleotide region containing a small stem loop and downstream interaction site (figure 1B) 87 (12). Due to the sequence similarity between the PKs (figure 1C), it is speculated that they 88 were formed by duplication events during viral replication, probably involving recombination. 89 Between two and four PKs are present in different virus isolates but no strain has been 90 identified with less than two PKs, emphasising their potential importance in the viral life cycle 91 (19, 20) . The presence of PKs has been reported in the 5′ UTR of other picornaviruses such as 92 author/funder. All rights reserved. No reuse allowed without permission. can occur in the absence of PKs at least one is required for wild-type (wt) replication. 104 Furthermore, competition experiments showed that extra copies of PKs conferred a replicative 105 advantage to genomes. Although replicons and full-length genomes lacking PKs were 106 replication-competent, no infectious virus was rescued from genomes containing less than one 107 PK copy. This is consistent with our earlier report describing the presence of putative 108 packaging signals in the PK region (22). 109 110 author/funder. All rights reserved. No reuse allowed without permission. Plasmid construction. 117 The FMDV replicon plasmids, pRep-ptGFP, and the replication-defective polymerase mutant 118 control, 3D-GNN, have already been described (10).\n",
      "\n",
      "120\n",
      "\n",
      "To introduce mutations into the PK region, the pRep-ptGFP replicon plasmid was digested 121 with SpeI and KpnI and the resulting fragment inserted into a sub-cloning vector (pBluescript) 122 to create the pBluescript PK. PKs 3 and 4 were removed by digestion with HindIII and AatII 123 before insertion of a synthetic DNA sequence with PK 3 and 4 deleted. PKs 2, 3 and 4 were 124 deleted by PCR amplification using ΔPK 234 Forward primer and FMDV 1331-1311 reverse 125 primer, the resultant product was digested with HindIII and AatII and ligated into the 126 pBluescript PK vector. Complete PK deletion was achieved by introduction of an AflII site at 127 the 3′ end of the poly-C tract by PCR mutagenesis to create the sub-cloning vector, pBluescript 128 C11, which was then used to remove all the PKs by PCR mutagenesis using ΔPK 1234 f\n"
     ]
    }
   ],
   "source": [
    "body = \"\"\n",
    "\n",
    "for section, text in texts_di.items():\n",
    "    body += section\n",
    "    body += \"\\n\\n\"\n",
    "    body += text\n",
    "    body += \"\\n\\n\"\n",
    "\n",
    "print(body[:3000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "VP3, and VP0 (which is further processed to VP2 and VP4 during virus assembly) (6). The P2 64 and P3 regions encode the non-structural proteins 2B and 2C and 3A, 3B (1-3) (VPg), 3C pro and 4 structural protein-coding region is replaced by reporter genes, allow the study of genome 68 replication without the requirement for high containment (9, 10) ( figure 1A ).\n",
      "\n",
      "70\n",
      "\n",
      "The FMDV 5′ UTR is the largest known picornavirus UTR, comprising approximately 1300 71 nucleotides and containing several highly structured regions. The first 360 nucleotides at the 5′ 72 end are predicted to fold into a single large stem loop termed the S-fragment, followed by a The PKs were originally predicted in 1987 and consist of two to four tandem repeats of a ~48 86 nucleotide region containing a small stem loop and downstream interaction site (figure 1B) 87 (12). Due to the sequence similarity between the PKs (figure 1C), it is speculated that they 88 were formed by duplication events during viral replication, probably involving recombination. 89 Between two and four PKs are present in different virus isolates but no strain has been 90 identified with less than two PKs, emphasising their potential importance in the viral life cycle 91 (19, 20) . The presence of PKs has been reported in the 5′ UTR of other picornaviruses such as 92 author/funder. All rights reserved. No reuse allowed without permission. can occur in the absence of PKs at least one is required for wild-type (wt) replication. 104 Furthermore, competition experiments showed that extra copies of PKs conferred a replicative 105 advantage to genomes. Although replicons and full-length genomes lacking PKs were 106 replication-competent, no infectious virus was rescued from genomes containing less than one 107 PK copy. This is consistent with our earlier report describing the presence of putative 108 packaging signals in the PK region (22). 109 110 author/funder. All rights reserved. No reuse allowed without permission. Plasmid construction. 117 The FMDV replicon plasmids, pRep-ptGFP, and the replication-defective polymerase mutant 118 control, 3D-GNN, have already been described (10).\n",
      "\n",
      "120\n",
      "\n",
      "To introduce mutations into the PK region, the pRep-ptGFP replicon plasmid was digested 121 with SpeI and KpnI and the resulting fragment inserted into a sub-cloning vector (pBluescript) 122 to create the pBluescript PK. PKs 3 and 4 were removed by digestion with HindIII and AatII 123 before insertion of a synthetic DNA sequence with PK 3 and 4 deleted. PKs 2, 3 and 4 were 124 deleted by PCR amplification using ΔPK 234 Forward primer and FMDV 1331-1311 reverse 125 primer, the resultant product was digested with HindIII and AatII and ligated into the 126 pBluescript PK vector. Complete PK deletion was achieved by introduction of an AflII site at 127 the 3′ end of the poly-C tract by PCR mutagenesis to create the sub-cloning vector, pBluescript 128 C11, which was then used to remove all the PKs by PCR mutagenesis using ΔPK 1234 f\n"
     ]
    }
   ],
   "source": [
    "print(format_body(file['body_text'])[:3000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['title', 'authors'])\n"
     ]
    }
   ],
   "source": [
    "# Biorxiv: Metadata\n",
    "print(all_files[0]['metadata'].keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The RNA pseudoknots in foot-and-mouth disease virus are dispensable for genome replication but essential for the production of infectious virus. 2 3\n"
     ]
    }
   ],
   "source": [
    "print(all_files[0]['metadata']['title'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'affiliation': {},\n",
      "  'email': '',\n",
      "  'first': 'Joseph',\n",
      "  'last': 'Ward',\n",
      "  'middle': ['C'],\n",
      "  'suffix': ''},\n",
      " {'affiliation': {},\n",
      "  'email': '',\n",
      "  'first': 'Lidia',\n",
      "  'last': 'Lasecka-Dykes',\n",
      "  'middle': [],\n",
      "  'suffix': ''},\n",
      " {'affiliation': {},\n",
      "  'email': '',\n",
      "  'first': 'Chris',\n",
      "  'last': 'Neil',\n",
      "  'middle': [],\n",
      "  'suffix': ''}]\n"
     ]
    }
   ],
   "source": [
    "authors = all_files[0]['metadata']['authors']\n",
    "pprint(authors[:3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: Joseph C Ward\n",
      "Affiliation: \n",
      "\n",
      "Name: Lidia Lasecka-Dykes\n",
      "Affiliation: \n",
      "\n",
      "Name: Chris Neil\n",
      "Affiliation: \n",
      "\n",
      "Name: Oluwapelumi Adeyemi\n",
      "Affiliation: \n",
      "\n",
      "Name: Sarah \n",
      "Affiliation: \n",
      "\n",
      "Name:  Gold\n",
      "Affiliation: \n",
      "\n",
      "Name: Niall Mclean\n",
      "Affiliation: \n",
      "\n",
      "Name: Caroline Wright\n",
      "Affiliation: \n",
      "\n",
      "Name: Morgan R Herod\n",
      "Affiliation: \n",
      "\n",
      "Name: David Kealy\n",
      "Affiliation: \n",
      "\n",
      "Name: Emma \n",
      "Affiliation: \n",
      "\n",
      "Name: Warner \n",
      "Affiliation: \n",
      "\n",
      "Name: Donald P King\n",
      "Affiliation: \n",
      "\n",
      "Name: Tobias J Tuthill\n",
      "Affiliation: \n",
      "\n",
      "Name: David J Rowlands\n",
      "Affiliation: \n",
      "\n",
      "Name: Nicola J \n",
      "Affiliation: \n",
      "\n",
      "Name: Stonehouse A#\n",
      "Affiliation: \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for author in authors:\n",
    "    print(\"Name:\", format_name(author))\n",
    "    print(\"Affiliation:\", format_affiliation(author['affiliation']))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'authors': [{'affiliation': {'institution': 'University of Georgia',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Salman',\n",
      "              'last': 'Butt',\n",
      "              'middle': ['L'],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'University of Georgia',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Eric',\n",
      "              'last': 'Erwood',\n",
      "              'middle': ['C'],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'University of Georgia',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Jian',\n",
      "              'last': 'Zhang',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'University of Georgia',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Holly',\n",
      "              'last': 'Sellers',\n",
      "              'middle': ['S'],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'University of Georgia',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Kelsey',\n",
      "              'last': 'Young',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'Virginia Polytechnical Institute '\n",
      "                                             'and State University',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Kevin',\n",
      "              'last': 'Lahmers',\n",
      "              'middle': ['K'],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'University of Georgia',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'James',\n",
      "              'last': 'Stanton',\n",
      "              'middle': ['B'],\n",
      "              'suffix': ''}],\n",
      " 'title': 'Real-time, MinION-based, amplicon sequencing for lineage typing of '\n",
      "          'infectious bronchitis virus from upper respiratory samples'}\n"
     ]
    }
   ],
   "source": [
    "pprint(all_files[4]['metadata'], depth=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formatting without affiliation:\n",
      "Salman L Butt, Eric C Erwood, Jian Zhang, Holly S Sellers, Kelsey Young, Kevin K Lahmers, James B Stanton\n",
      "\n",
      "Formatting with affiliation:\n",
      "Salman L Butt (University of Georgia, 30602, Athens, GA, USA), Eric C Erwood (University of Georgia, 30602, Athens, GA, USA), Jian Zhang (University of Georgia, 30602, Athens, GA, USA), Holly S Sellers (University of Georgia, 30602, Athens, GA, USA), Kelsey Young (University of Georgia, 30602, Athens, GA, USA), Kevin K Lahmers (Virginia Polytechnical Institute and State University, 24061, Blacksburg, VA, USA), James B Stanton (University of Georgia, 30602, Athens, GA, USA)\n"
     ]
    }
   ],
   "source": [
    "authors = all_files[4]['metadata']['authors']\n",
    "print(\"Formatting without affiliation:\")\n",
    "print(format_authors(authors, with_affiliation=False))\n",
    "print(\"\\nFormatting with affiliation:\")\n",
    "print(format_authors(authors, with_affiliation=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'authors': [{'first': 'T', 'last': 'Jackson', 'middle': [], 'suffix': ''},\n",
      "              {'first': 'T', 'last': 'Tuthill', 'middle': [...], 'suffix': ''},\n",
      "              {'first': 'D', 'last': 'Rowlands', 'middle': [...], 'suffix': ''},\n",
      "              {'first': 'N',\n",
      "               'last': 'Stonehouse',\n",
      "               'middle': [...],\n",
      "               'suffix': ''}],\n",
      "  'issn': '',\n",
      "  'other_ids': {},\n",
      "  'pages': '',\n",
      "  'ref_id': 'b0',\n",
      "  'title': 'Genetic economy in 598 picornaviruses: Foot-and-mouth disease '\n",
      "           'virus replication exploits alternative precursor 599 cleavage '\n",
      "           'pathways',\n",
      "  'venue': 'PLOS Pathog',\n",
      "  'volume': '13',\n",
      "  'year': 2017},\n",
      " {'authors': [{'first': 'N',\n",
      "               'last': 'Sanderson',\n",
      "               'middle': [...],\n",
      "               'suffix': ''},\n",
      "              {'first': 'N', 'last': 'Knowles', 'middle': [...], 'suffix': ''},\n",
      "              {'first': 'D', 'last': 'King', 'middle': [...], 'suffix': ''},\n",
      "              {'first': 'E', 'last': 'Cottam', 'middle': [...], 'suffix': ''}],\n",
      "  'issn': '',\n",
      "  'other_ids': {},\n",
      "  'pages': '',\n",
      "  'ref_id': 'b2',\n",
      "  'title': 'A universal protocol to 602 generate consensus level genome '\n",
      "           'sequences for foot-and-mouth disease virus and other 603 '\n",
      "           'positive-sense polyadenylated RNA viruses using the Illumina MiSeq',\n",
      "  'venue': 'BMC Genomics',\n",
      "  'volume': '604',\n",
      "  'year': 2014}]\n"
     ]
    }
   ],
   "source": [
    "# Biorxiv: bibliography\n",
    "bibs = list(file['bib_entries'].values())\n",
    "pprint(bibs[:2], depth=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'N D Sanderson, N J Knowles, D P King, E M Cottam'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "format_authors(bibs[1]['authors'], with_affiliation=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genetic economy in 598 picornaviruses: Foot-and-mouth disease virus replication exploits alternative precursor 599 cleavage pathways, T Jackson, T J Tuthill, D J Rowlands, N J Stonehouse, PLOS Pathog, 2017; A universal protocol to 602 generate consensus level genome sequences for foot-and-mouth disease virus and other 603 positive-sense polyadenylated RNA viruses using the Illumina MiSeq, N D Sanderson, N J Knowles, D P King, E M Cottam, BMC Genomics, 2014; Library preparation for highly accurate population 606 sequencing of RNA viruses, A Acevedo, R Andino, Nat Protoc, 2014; IDBA-UD: a de novo assembler for 608 single-cell and metagenomic sequencing data with highly uneven depth, Y Peng, Hcm Leung, S M Yiu, Fyl Chin, , 2012; Basic local alignment 611 search tool, S F Altschul, W Gish, W Miller, E W Myers, D J Lipman, J Mol Biol, 1990\n"
     ]
    }
   ],
   "source": [
    "bib_formatted = format_bib(bibs[:5])\n",
    "print(bib_formatted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7b17633e2bdf448fbbe58d3e165da5e9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=1625.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Biorxiv: Generate CSV\n",
    "cleaned_files = []\n",
    "\n",
    "for file in tqdm(all_files):\n",
    "    features = [\n",
    "        file['paper_id'],\n",
    "        file['metadata']['title'],\n",
    "        format_authors(file['metadata']['authors']),\n",
    "        format_authors(file['metadata']['authors'], \n",
    "                       with_affiliation=True),\n",
    "        format_body(file['abstract']),\n",
    "        format_body(file['body_text']),\n",
    "        format_bib(file['bib_entries']),\n",
    "        file['metadata']['authors'],\n",
    "        file['bib_entries']\n",
    "    ]\n",
    "    \n",
    "    cleaned_files.append(features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>affiliations</th>\n",
       "      <th>abstract</th>\n",
       "      <th>text</th>\n",
       "      <th>bibliography</th>\n",
       "      <th>raw_authors</th>\n",
       "      <th>raw_bibliography</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0015023cc06b5362d332b3baf348d11567ca2fbb</td>\n",
       "      <td>The RNA pseudoknots in foot-and-mouth disease ...</td>\n",
       "      <td>Joseph C Ward, Lidia Lasecka-Dykes, Chris Neil...</td>\n",
       "      <td>Joseph C Ward, Lidia Lasecka-Dykes, Chris Neil...</td>\n",
       "      <td>Abstract\\n\\nword count: 194 22 Text word count...</td>\n",
       "      <td>\\n\\nVP3, and VP0 (which is further processed t...</td>\n",
       "      <td>Genetic economy in 598 picornaviruses: Foot-an...</td>\n",
       "      <td>[{'first': 'Joseph', 'middle': ['C'], 'last': ...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Genetic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>00340eea543336d54adda18236424de6a5e91c9d</td>\n",
       "      <td>Analysis Title: Regaining perspective on SARS-...</td>\n",
       "      <td>Carla Mavian, Simone Marini, Costanza Manes, I...</td>\n",
       "      <td>Carla Mavian (University of Florida, Gainesvil...</td>\n",
       "      <td>Abstract\\n\\nDuring the past three months, a ne...</td>\n",
       "      <td>\\n\\nIn December 2019, a novel coronavirus, SAR...</td>\n",
       "      <td>Situation Report -43, , Coronavirus disease 20...</td>\n",
       "      <td>[{'first': 'Carla', 'middle': [], 'last': 'Mav...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Situati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>004f0f8bb66cf446678dc13cf2701feec4f36d76</td>\n",
       "      <td>Healthcare-resource-adjusted vulnerabilities t...</td>\n",
       "      <td>Hanchu Zhou, Jiannan Yang, Kaicheng Tang, † , ...</td>\n",
       "      <td>Hanchu Zhou (City University of Hong Kong, Hon...</td>\n",
       "      <td></td>\n",
       "      <td>Introduction\\n\\nThe 2019-nCoV epidemic has spr...</td>\n",
       "      <td>World Health Organizations. Novel Coronavirus ...</td>\n",
       "      <td>[{'first': 'Hanchu', 'middle': [], 'last': 'Zh...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'World H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>00911cf4f99a3d5ae5e5b787675646a743574496</td>\n",
       "      <td>CHEER: hierarCHical taxonomic classification f...</td>\n",
       "      <td>Jiayu Shang, Yanni Sun</td>\n",
       "      <td>Jiayu Shang (City University of Hong Kong, Hon...</td>\n",
       "      <td>Abstract\\n\\nThe fast accumulation of viral met...</td>\n",
       "      <td>Introduction\\n\\nMetagenomic sequencing, which ...</td>\n",
       "      <td>Application of metagenomics in the human gut m...</td>\n",
       "      <td>[{'first': 'Jiayu', 'middle': [], 'last': 'Sha...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Applica...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>00d16927588fb04d4be0e6b269fc02f0d3c2aa7b</td>\n",
       "      <td>Real-time, MinION-based, amplicon sequencing f...</td>\n",
       "      <td>Salman L Butt, Eric C Erwood, Jian Zhang, Holl...</td>\n",
       "      <td>Salman L Butt (University of Georgia, 30602, A...</td>\n",
       "      <td>Abstract\\n\\nInfectious bronchitis (IB) causes ...</td>\n",
       "      <td>Introduction\\n\\nInfectious bronchitis (IB), wh...</td>\n",
       "      <td>Emergence of novel strains of avian infectious...</td>\n",
       "      <td>[{'first': 'Salman', 'middle': ['L'], 'last': ...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Emergen...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  0015023cc06b5362d332b3baf348d11567ca2fbb   \n",
       "1  00340eea543336d54adda18236424de6a5e91c9d   \n",
       "2  004f0f8bb66cf446678dc13cf2701feec4f36d76   \n",
       "3  00911cf4f99a3d5ae5e5b787675646a743574496   \n",
       "4  00d16927588fb04d4be0e6b269fc02f0d3c2aa7b   \n",
       "\n",
       "                                               title  \\\n",
       "0  The RNA pseudoknots in foot-and-mouth disease ...   \n",
       "1  Analysis Title: Regaining perspective on SARS-...   \n",
       "2  Healthcare-resource-adjusted vulnerabilities t...   \n",
       "3  CHEER: hierarCHical taxonomic classification f...   \n",
       "4  Real-time, MinION-based, amplicon sequencing f...   \n",
       "\n",
       "                                             authors  \\\n",
       "0  Joseph C Ward, Lidia Lasecka-Dykes, Chris Neil...   \n",
       "1  Carla Mavian, Simone Marini, Costanza Manes, I...   \n",
       "2  Hanchu Zhou, Jiannan Yang, Kaicheng Tang, † , ...   \n",
       "3                             Jiayu Shang, Yanni Sun   \n",
       "4  Salman L Butt, Eric C Erwood, Jian Zhang, Holl...   \n",
       "\n",
       "                                        affiliations  \\\n",
       "0  Joseph C Ward, Lidia Lasecka-Dykes, Chris Neil...   \n",
       "1  Carla Mavian (University of Florida, Gainesvil...   \n",
       "2  Hanchu Zhou (City University of Hong Kong, Hon...   \n",
       "3  Jiayu Shang (City University of Hong Kong, Hon...   \n",
       "4  Salman L Butt (University of Georgia, 30602, A...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Abstract\\n\\nword count: 194 22 Text word count...   \n",
       "1  Abstract\\n\\nDuring the past three months, a ne...   \n",
       "2                                                      \n",
       "3  Abstract\\n\\nThe fast accumulation of viral met...   \n",
       "4  Abstract\\n\\nInfectious bronchitis (IB) causes ...   \n",
       "\n",
       "                                                text  \\\n",
       "0  \\n\\nVP3, and VP0 (which is further processed t...   \n",
       "1  \\n\\nIn December 2019, a novel coronavirus, SAR...   \n",
       "2  Introduction\\n\\nThe 2019-nCoV epidemic has spr...   \n",
       "3  Introduction\\n\\nMetagenomic sequencing, which ...   \n",
       "4  Introduction\\n\\nInfectious bronchitis (IB), wh...   \n",
       "\n",
       "                                        bibliography  \\\n",
       "0  Genetic economy in 598 picornaviruses: Foot-an...   \n",
       "1  Situation Report -43, , Coronavirus disease 20...   \n",
       "2  World Health Organizations. Novel Coronavirus ...   \n",
       "3  Application of metagenomics in the human gut m...   \n",
       "4  Emergence of novel strains of avian infectious...   \n",
       "\n",
       "                                         raw_authors  \\\n",
       "0  [{'first': 'Joseph', 'middle': ['C'], 'last': ...   \n",
       "1  [{'first': 'Carla', 'middle': [], 'last': 'Mav...   \n",
       "2  [{'first': 'Hanchu', 'middle': [], 'last': 'Zh...   \n",
       "3  [{'first': 'Jiayu', 'middle': [], 'last': 'Sha...   \n",
       "4  [{'first': 'Salman', 'middle': ['L'], 'last': ...   \n",
       "\n",
       "                                    raw_bibliography  \n",
       "0  {'BIBREF0': {'ref_id': 'b0', 'title': 'Genetic...  \n",
       "1  {'BIBREF0': {'ref_id': 'b0', 'title': 'Situati...  \n",
       "2  {'BIBREF0': {'ref_id': 'b0', 'title': 'World H...  \n",
       "3  {'BIBREF0': {'ref_id': 'b0', 'title': 'Applica...  \n",
       "4  {'BIBREF0': {'ref_id': 'b0', 'title': 'Emergen...  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_names = [\n",
    "    'paper_id', \n",
    "    'title', \n",
    "    'authors',\n",
    "    'affiliations', \n",
    "    'abstract', \n",
    "    'text', \n",
    "    'bibliography',\n",
    "    'raw_authors',\n",
    "    'raw_bibliography'\n",
    "]\n",
    "\n",
    "clean_df = pd.DataFrame(cleaned_files, columns=col_names)\n",
    "clean_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_df.to_csv('biorxiv_clean.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b73704c45ba34530b982f722dfdc630f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=26505.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate CSV: Custom (PMC), Commercial, Non-commercial licenses\n",
    "pmc_dir = 'CORD-19-research-challenge/custom_license/custom_license/pdf_json/'\n",
    "pmc_files = load_files(pmc_dir)\n",
    "pmc_df = generate_clean_df(pmc_files)\n",
    "pmc_df.to_csv('clean_pmc.csv', index=False)\n",
    "pmc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comm_dir = 'CORD-19-research-challenge/comm_use_subset/comm_use_subset/pdf_json/'\n",
    "comm_files = load_files(comm_dir)\n",
    "comm_df = generate_clean_df(comm_files)\n",
    "comm_df.to_csv('clean_comm_use.csv', index=False)\n",
    "comm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "noncomm_dir = 'CORD-19-research-challenge/noncomm_use_subset/noncomm_use_subset/pdf_json/'\n",
    "noncomm_files = load_files(noncomm_dir)\n",
    "noncomm_df = generate_clean_df(noncomm_files)\n",
    "noncomm_df.to_csv('clean_noncomm_use.csv', index=False)\n",
    "noncomm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
